### DISCUSSION TITLE
Add countof(SomeEnum) as a compile-time constant

### DISCUSSION THREAD

**Discussion by Entiat** [dotnet/csharplang#8906] [Language Ideas]
_(I have zero doubt this has come up before, multiple times, I've searched but no luck - my apologies if the reason this was never implemented is "common knowledge" and I've just missed it)_

There is currently no way to determine the count of an enumerated type as a compile-time constant. This means this count cannot be used in many useful scenarios, including default method parameter values, array sizes at compile time, and as attribute parameters. 

Workarounds in wide use are klunky and error-prone.

Example of use:

```
public enum MyEnum
{
     FirstItem,
     SecondItem,
     ThirdItem
}

// Later...

[AttributeWithAnIntParameter(countof(MyEnum))]
public SomeMethod(int size = countof(MyEnum))
{
     // code that uses size...
}

```
Examples of a couple error-prone workarounds in current common usage:

```
public enum MyFirstEnum
{
     FirstItem,
     SecondItem,
     ThirdItem
}
public const int MyFirstEnumCount = 3;

// Some relevant static class somewhere
static RelevantClass()
{
    Debug.Assert(MyFirstEnumCount == Enum.GetNames(typeof(MyFirstEnum)).Length, "MyFirstEnumCount is out of sync with the actual count of enum items");
}
```
Or:
```
public enum MySecondEnum
{
     FirstItem,
     SecondItem = 42,
     ThirdItem = 0x1234,

     countof = 3
}

// Some relevant static class somewhere
static RelevantClass()
{
    Debug.Assert(MySecondEnum.count == Enum.GetNames(typeof(MySecondEnum)).Length, "MySecondEnum.countof value is out of sync with actual count of enum items");
}
```

**Comment by HaloFour** [comment]
A compile-time constant wouldn't work well if the `enum` was defined in another assembly and the number of items could change.  Then then count would be wrong which could negatively affect code.  Better to calculate the length once at runtime, which is cheap and has the advantage of being correct.

**Reply by colejohnson66** [reply]
Perhaps a runtime solution? `Enum.GetCount`?

**Reply by mysticetus** [reply]
If you change the values of an enum in a referenced assembly, that will already break any references to them. My (admittedly naive) understanding is the compiler+runtime has various hash codes, versioning, and other systems in place to figure that out and fire errors, refuse to run, etc. Again, maybe it's a lack of deep knowledge on my part (I'm not a compiler developer), but I'm having trouble seeing why adding "count" via whatever mechanism names+values are exposed would make matters worse. (?)

**Reply by CyrusNajmabadi** [reply]
> that will already break any references to them. 

Yes.  If you change the value.  But not the count.  Each individual value stays a constant.  But it's never been an aspect of hte .net ecosystem that the *count* of elements is an unchanging constant.

**Reply by HaloFour** [reply]
> If you change the values of an enum in a referenced assembly, that will already break any references to them.

Only if you would do so in a way that would cause them to be renumbered.  Adding a new value to an enum is not a binary breaking change and the runtime has added new values to some enums on multiple occasions.

**Reply by mysticetus** [reply]
It hasn't been an aspect of the ecosystem, so far :-)

What are the downsides of making such an update to the language? I 100% understand language advancement involves weighing complex pros and cons. If countof(enumType) was added, and fully supported - what problems would that cause? Would it turn into a rabbit hole of cascading dependency problems? Or...are there other fundamental issues I'm blissfully ignorant of?

On the pro side, it would eliminate a certain category of bugs and kludgy code. Is that outweighed by the cons? (seriously asking - my knowledge isn't deep enough to have a valid opinion at that level)

**Reply by CyrusNajmabadi** [reply]
> On the pro side, it would eliminate a certain category of bugs and kludgy code.

What examples are tehre of this?  I'm not understanding the use cases here.

**Reply by Entiat** [reply]
See my original post. Default method parameter values and attribute parameters are my current stumbling blocks. Apparently the Range attribute in Unity is a source of consternation for at least a couple other people on this front, fwiw - my issues with attribute parameters involve my own custom attributes, which means (for me at least) I can re-jigger them to figure things out via other compile time mechanisms, but it's not ideal and triggers my "that looks like a bug waiting to happen" spidey-sense.

I also show some of the common workarounds. I personally use the "CountOf = 3" with runtime assertion approach to detect count becoming out of syn - which has obvious problems if, say, the static constructor ends up not being called due to other code changes.

Searching SO and other resources show some...not so wonderful...workarounds involving reflection as well.

**Reply by Entiat** [reply]
And for the record - I accidentally posted some responses from a work account (mysticetus) above - doh. That's actually me.

(Not trying to make it look like there's increased attention to this issue. I just used the wrong browser.)

This proposal isn't life or death, but would be nice to improve one aspect of the language that's annoying for certain use cases, and has the potential to cause some specific, possibly hard-to-find bugs. 

Thanks for taking the time to invest some brain cycles on it :-)

**Reply by CyrusNajmabadi** [reply]
> See my original post. Default method parameter values

But why does the default parameter value need to know the count of enum values?  I'm still not understanding that part. :)

Note: this is a perfect example of what would break.  A value is added to the enum, and all downstream code is broken because it hardcodes in the value of hte defualt-parameter-value at the *callsite* at *compile time*.  So unless you recompile downstream dependencies they'll now be broken :)

**Reply by jnm2** [reply]
> Perhaps a runtime solution? `Enum.GetCount`?

`Enum.GetValues(typeof(XyzEnum)).Length` works for this.

It does raise a question about what you want for `enum XyzEnum { A = 1, AliasForA = 1 }`.

**Reply by Entiat** [reply]
Current use cases for me are twofold:

**_First, a type safe (ish) way to index arrays_**. For example, I am currently building out a marine mammal species name translation class that knows about common English, common Spanish, Latin (scienfitic), and a weird, archaic code system for marine mammals that some biologist invented when the dinosaurs roamed the earth...well, the 1970s :-) but is still in use by a few biologists today.

Since the numbers are relatively small (~130 marine mammal species) using an array of arrays works better (fewer bugs, my use case is fine with in-order traversal pattern matching) than more complex indexed/hashed data structures one might consider for general purpose translation APIs.<br>

Anyways, the solution path I've chosen is to define an enum to facilitate language selection:
```
public enum Language : int
{
     CommonEnglish,
     CommonSpanish,
     Latin,
     KenneyCode,

     CountOf = 4
}
```
This then facilitates allocation of 2D arrays of names that can be iterated, translated, indexed via ToLower and a Dictionary if it becomes necessary, etc. Adding another language is a simple matter of adding it to the enum (and hopefully remembering to update CountOf) and then dealing with compilation errors because things break at compile time if I forgot to add something.

I understand there are runtime workarounds (for example: Enum.GetValues(type).Length in static constructors) for this and I use them. They are, however, a bit klunkier and require I never pass these lengths as default parameters to initialization methods. More importantly, instead of finding my own coding errors around these constructs at compile time, I (hopefully!) need to find them during runtime testing (Debug.Asserts everywhere) before shipping to my customers.

**_Second, Parameters for attributes_**. I've mentioned others out there using Unity's Range Attribute (search SO for examples) which takes integers and it can be useful to leverage the count of an enumerated type in certain circumstances. My use cases are similar, albeit relative to UI for recording data about marine life and my Attribute-derived classes want to know the count of an enum in a constructor parameter. 

Similar problem space, with similar workarounds and similar (not horrifyingly bad, but still non-zero) increased bug risks - finding bugs is based on Debug.Assert and exceptions at runtime rather than compile-time errors.

-----

As I said, this issue isn't the end of the world - but if it were implemented in a constrained way (literally just the count of items, maybe define it to match what Enum.GetValues(type).Length returns at runtime?) would be nice and could reduce the frequency of a certain set of bugs.

**Reply by CyrusNajmabadi** [reply]
> First, a type safe (ish) way to index arrays

This use case makes sense. But in that case, the sensible operation to me is getting the *max* enum value. That way you can create an array one larger than that.  Then you can index safely given any of the current values in the enum. 

Similarly, an operation that does the same, but is the max flags value, again for indexing. 

'Count' doesn't seem right here.

**Reply by CyrusNajmabadi** [reply]
> I understand there are runtime workarounds (for example: Enum.GetValues(type).Length in static constructors) for this and I use them

Note: given your use cases, I don't view this as a workaround. It seems like the only viable path.  This should not be a constant here. Otherwise, if you add a single member, your downstairs built binaries will break (since they will now allocate an array that is too small for the number of named enum members they can receive. 

This is truly not a constant value, so even for your use cases you'd want and need this to be computed at runtime for correctness.

**Reply by mysticetus** [reply]
How is this different than deleting an enum value, which is also a breaking change for downstream dependent assemblies? Or any other breaking change (symbol name change, class declaration change, etc.)

These then cause the downstream assemblies to be rebuilt (if that's what you are doing), and if they reference the deleted enum a compiler error is generated. I'm not an expert on what happens to dependent assemblies if you *don't* rebuild them (I don't do that with my code) - but I assume there are versioning/hash/checksum checks in place to detect this and fail to load (or something?). 

Is there a reason we can't do the same dependency-tracking and rebuild/error-on-load for countof()?

**Reply by CyrusNajmabadi** [reply]
> How is this different than deleting an enum value, which is also a breaking change for downstream dependent assemblies? Or any other breaking change (symbol name change, class declaration change, etc.)

Because adding an enum value has never been a breaking change. It's different because it's not at all like any of those things.

**Comment by jnm2** [comment]
This is another case (similar to https://github.com/dotnet/csharplang/issues/8505, `typeof` string constants) where you can use a source generator today to declare the constant that you want to use.
